Add RSpec.current_scope method to replace `currently_executing_a_context_hook?` and `self.inspect` hack

odinhb opened the pull request:
fixes #2893

pirj commented on lib/rspec/core/example.rb:
```diff
@@ -248,6 +248,8 @@ def run(example_group_instance, reporter)
         RSpec.configuration.configure_example(self, hooks)
         RSpec.current_example = self
+        prev_scope = RSpec.current_scope
+        RSpec.current_scope = :example
```
It seems to me this should go inside `with_around_and_singleton_context_hooks` block.

odinhb replied:
It ended up inside `with_around_example_hooks` which is called inside `with_around_and_singleton_context_hooks` thanks to your suggestion.

pirj commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,49 @@
+    end
+
+    it "works for more than one example in a describe block" do
+      expect(RSpec.current_scope).to eq(:example)
```
Wondering if all the above expectations can be made inside a synthetic example group, i.e.:
```ruby
it "..." do
  group = RSpec.describe
  group.before(:context) { expect(RSpec.current_scope).to be(:example_group) }
  # etc
end
```

odinhb replied:
This is possible, but I'm not getting very good feedback when my expectations fail. I'd like to see exactly where it is not yet implemented correctly when it fails.

Synthetic example groups are giving me unrelated errors (from `it_behaves_like 'library wide checks', 'rspec-core'`, i think?).

Also, I found specs for the old implementation. It has some oversights:
```rb
it "does not set currently_executing_a_context_hook? outside of before(:context) execution" do
  group = RSpec.describe
  hook_result = nil

  group.before(:context) { hook_result = group.currently_executing_a_context_hook? }
  # Notice how the method is called on the group assigned to a variable.
  # If the example was realistic it would be called on the value of self and the hook would raise.
  group.before(:each) { hook_result = group.currently_executing_a_context_hook? }
  # We also forgot to test what the method does inside an example... Where it also raises
  group.example("") {}
  group.run
  expect(hook_result).to be false
end
```

So I think it'd be better to dogfood/integration test it.

Would these specs be more appropriately located in the spec/integration folder?
Should I write a .feature too? (which would also help document this things existence)

odinhb replied:
I wrote a cucumber spec for it. Turned out to be the best fit since messing with global configuration caused unrelated failures in the rspec tests.

I still did not find room to use synthetic example groups, since they need to look like this:

```rb
it "bla" do
  yays = []
  group = RSpec.describe
  group.before(:context) do
    expect(RSpec.current_scope).to be(:before_context_hook)
    yays << :bc
  end
  group.after(:context) do
    expect(RSpec.current_scope).to eq(:after_context_hook)
    yays << :ac
  end

  group.it "be like this" do
    expect(RSpec.current_scope).to eq(:example)
    yays << :ex
  end

  group.run

  expect(yays).to eq([:bc, :ex, :ac])
end
```

Which is a bit more cumbersome than I'd like when writing/reading/fixing specs.

pirj commented:
Nice.
Actually, now when I look at it, it seems to me there is yet another scope, but I don't have a good term to describe it.

```ruby
RSpec.describe Thing, :type => :model do

  # scope: example_group, context: example group class
  fixtures :things
  things(:one) # BOOM
  let_it_be(:stuff) { ... }
  stuff # BOOM

  before(:context) do
    # scope: example_group_hook, context: example group instance?
    things(:one) # BOOM
    stuff # BOOM    
  end

  it "fixture method defined" do
    # scope: example, context: example group instance
    # same as inside `before(:example)` for the needs of this PR
    things(:one) # OK
    stuff # OK
  end
end
```

We need to differentiate between the latter two, is that correct?

JonRowe commented on lib/rspec/core.rb:
```diff
@@ -129,6 +129,20 @@ def self.current_example=(example)
 
+  # Get the current scope rspec is currently executing in
+  # Returns nil until it starts executing an example_group
+  # TODO ... something something actually what it returns nil when i call it in an example group?
```
The first execution could set this to `:suite` and similarly reset it to `:suite` before the "after suite" hooks are run.

odinhb replied:
Went w/ setting it to :suite as soon as rspec loads, and then setting it to nil after the suite and all hooks are done.

JonRowe commented on lib/rspec/core/example_group.rb:
```diff
@@ -598,6 +598,8 @@ def self.run_after_context_hooks(example_group_instance)
       def self.run(reporter=RSpec::Core::NullReporter)
         return if RSpec.world.wants_to_quit
+        prev_scope = RSpec.current_scope
+        RSpec.current_scope = :example_group
```
Similarly, I'd set it back at line 607, and set it to example at 606

odinhb replied:
I don't think this is relevant anymore. Also I'm not quite following.

JonRowe commented on lib/rspec/core.rb:
```diff
@@ -129,6 +129,30 @@ def self.current_example=(example)
+  #
+  # Returns `:suite` while loading your tests
+  #
+  # Returns `:before_each_hook`/`:before_all_hook`/`:before_suite_hook`/`:after_each_hook`/`:after_all_hook`/`:after_suite_hook`
```
`s/each/example` `s/all/context/`

JonRowe replied:
```suggestion
  # Returns `:before_example_hook`/`:before_context_hook`/`:before_suite_hook`/`:after_example_hook`/`:after_context_hook`/`:after_suite_hook`
```

JonRowe commented on lib/rspec/core/example.rb:
```diff
@@ -278,6 +281,7 @@ def run(example_group_instance, reporter)
               rescue AllExceptionsExcludingDangerousOnesOnRubiesThatAllowIt => e
                 set_exception(e)
               ensure
+                RSpec.current_scope = :after_each_hook
```
```suggestion
                RSpec.current_scope = :after_example_hook
```

JonRowe commented on lib/rspec/core.rb:
```diff
@@ -129,6 +129,30 @@ def self.current_example=(example)
+  # Returns `:before_each_hook`/`:before_all_hook`/`:before_suite_hook`/`:after_each_hook`/`:after_all_hook`/`:after_suite_hook`
+  # in the respective hooks
+  #
+  # Returns `before_each_hook`/`after_each_hook` inside an `around :each` hook, before an after you call `example.run` respectively.
```
```suggestion
  # Returns `before_example_hook`/`after_example_hook` inside an `around :example` hook, before an after you call `example.run` respectively.
```

JonRowe commented on lib/rspec/core/example.rb:
```diff
@@ -462,6 +468,7 @@ def hooks
       end
 
       def with_around_example_hooks
+        RSpec.current_scope = :before_each_hook
```
```suggestion
        RSpec.current_scope = :before_example_hook
```

JonRowe commented on lib/rspec/core/example_group.rb:
```diff
@@ -598,10 +598,12 @@ def self.run_after_context_hooks(example_group_instance)
 
         should_run_context_hooks = descendant_filtered_examples.any?
         begin
+          RSpec.current_scope = :before_all_hook
```
```suggestion
          RSpec.current_scope = :before_context_hook
```

JonRowe commented on lib/rspec/core/example_group.rb:
```diff
@@ -614,8 +616,10 @@ def self.run(reporter=RSpec::Core::NullReporter)
           RSpec.world.wants_to_quit = true if reporter.fail_fast_limit_met?
           false
         ensure
+          RSpec.current_scope = :after_all_hook
```
```suggestion
          RSpec.current_scope = :after_context_hook
```

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,74 @@
+        expect(RSpec.current_scope).to eq(:before_suite_hook)
+      end
+
+      c.before :all do
+        expect(RSpec.current_scope).to eq(:before_all_hook)
```
```suggestion
      c.before :context do
        expect(RSpec.current_scope).to eq(:before_context_hook)
```

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,74 @@
+        expect(RSpec.current_scope).to eq(:before_all_hook)
+      end
+
+      c.before :each do
+        expect(RSpec.current_scope).to eq(:before_each_hook)
```
```suggestion
      c.before :example do
        expect(RSpec.current_scope).to eq(:before_example_hook)
```

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,74 @@
+        expect(RSpec.current_scope).to eq(:before_each_hook)
+      end
+
+      c.around :each do |ex|
+        expect(RSpec.current_scope).to eq(:before_each_hook)
+        ex.run
+        expect(RSpec.current_scope).to eq(:after_each_hook)
```
```suggestion
      c.around :example do |ex|
        expect(RSpec.current_scope).to eq(:before_example_hook)
        ex.run
        expect(RSpec.current_scope).to eq(:after_example_hook)
```

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,74 @@
+        expect(RSpec.current_scope).to eq(:after_each_hook)
+      end
+
+      c.after :each do
+        expect(RSpec.current_scope).to eq(:after_each_hook)
```
```suggestion
      c.after :example do
        expect(RSpec.current_scope).to eq(:after_example_hook)
```

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,74 @@
+        expect(RSpec.current_scope).to eq(:after_each_hook)
+      end
+
+      c.after :all do
+        expect(RSpec.current_scope).to eq(:after_all_hook)
```
```suggestion
      c.after :context do
        expect(RSpec.current_scope).to eq(:after_context_hook)
```

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,74 @@
+      end
+    end
+
+    before :all do
+      expect(RSpec.current_scope).to eq(:before_all_hook)
+    end
+
+    before :each do
+      expect(RSpec.current_scope).to eq(:before_each_hook)
+    end
+
+    around :each do |ex|
+      expect(RSpec.current_scope).to eq(:before_each_hook)
+      ex.run
+      expect(RSpec.current_scope).to eq(:after_each_hook)
+    end
+
+    after :each do
+      expect(RSpec.current_scope).to eq(:after_each_hook)
+    end
+
+    after :all do
+      expect(RSpec.current_scope).to eq(:after_all_hook)
+    end
```
```suggestion
    before :context do
      expect(RSpec.current_scope).to eq(:before_context_hook)
    end

    before :example do
      expect(RSpec.current_scope).to eq(:before_example_hook)
    end

    around :example do |ex|
      expect(RSpec.current_scope).to eq(:before_example_hook)
      ex.run
      expect(RSpec.current_scope).to eq(:after_example_hook)
    end

    after :example do
      expect(RSpec.current_scope).to eq(:after_example_hook)
    end

    after :context do
      expect(RSpec.current_scope).to eq(:after_context_hook)
    end
```

JonRowe reviewed (commented):
The correct names for hooks are `:example` `:context` and `:suite`, they are aliased to `:each` and `:all` but are not the main names, so that should reflect that.

I also feel we shouldn't be restoring previous states but explicitly setting the state we are moving to, otherwise a bug will persist through that stack

odinhb commented:
> The correct names for hooks are `:example` `:context` and `:suite`, they are aliased to `:each` and `:all` but are not the main names, so that should reflect that.
> 
> I also feel we shouldn't be restoring previous states but explicitly setting the state we are moving to, otherwise a bug will persist through that stack

Fixed and fixed. Turns out the prev_scope thing was unnecessary anyway.

odinhb commented:
Did another iteration on this and I'm starting to feel satisfied with the implementation. Let me know if I missed anything else.

odinhb commented:
I'm having some real trouble with rubocop here. Pipeline says "74 files inspected, 1 offense detected", but my local install says "74 files inspected, 94 offenses detected". I can't figure out the difference between our setups.

JonRowe commented on lib/rspec/core/example_group.rb:
```diff
@@ -537,9 +537,15 @@ def self.store_before_context_ivars(example_group_instance)
         end
       end
 
+      # @deprecated use `RSpec.current_scope` instead
       # Returns true if a `before(:context)` or `after(:context)`
       # hook is currently executing.
       def self.currently_executing_a_context_hook?
+        RSpec.deprecate(
+          "currently_executing_a_context_hook",
+          :replacement => "RSpec.current_scope",
+        )
+
```
Do we need to deprecate this?

pirj replied:
I couldn't find any usages. Apparently, we planned to make use of it in `rspec-rails`, but didn't.
We could get rid of it in 4.0, and deprecate it in 3.x.

odinhb replied:
This is an old (and rather ineffective) implementation that I found, trying to solve the same problem as the new implementation.

Should I do anything special or are you guys gonna take care of the difference between 3.x/4.0?

pirj replied:
You're totally welcome to remove this now deprecated method by sending a PR against `4-0-dev` branch once this one is merged.

pirj replied:
A pull request to `rspec-rails` is welcome, too.

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -256,7 +270,7 @@ def reporter
       RSpec.clear_examples
 
-      RSpec.configuration.deprecation_stream = StringIO.new(deprecations = "".dup)
+      RSpec.configuration.deprecation_stream = StringIO.new(deprecations = +"")
```
Can you explain this?

pirj replied:
This would break pre-2.0 builds. We still maintain 1.8.7+ compatibility in RSpec 3.x.

odinhb replied:
Yeah my rubocop (which is wrong) told me to do it. Will revert.

odinhb replied:
```
rubocop: warning
Performance/UnfreezeString - Use unary plus to get an unfrozen string literal.
```

But there is no magic comment nor are we in a ruby version w/ frozen strings by default so why? Idk.

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,20 @@
   end
 
+  describe ".current_scope" do
+    before :all do
```
```suggestion
    before :context do
```

JonRowe commented on spec/rspec/core_spec.rb:
```diff
@@ -129,6 +129,20 @@
+      expect(RSpec.current_scope).to eq(:before_context_hook)
+    end
+
+    before :each do
```
```suggestion
    before do
```

pirj commented on lib/rspec/core.rb:
```diff
@@ -129,6 +129,31 @@ def self.current_example=(example)
+  # Returns `:example` inside it/example blocks
+  #
+  # Returns `nil` after your suite and all hooks are done
+  # @return [Symbol/Nil]
```
I believe writing `[Symbol, nil]` or `[Symbol, NilClass]` is more correct.

pirj commented on spec/rspec/core_spec.rb:
```diff
@@ -325,4 +339,3 @@ def reporter
     expect(status.exitstatus).to eq(0)
   end
 end
```
Please revert unrelated changes. We'll have to deal with merging a growing list of changes from 3.x branch to 4.0, that removed many things, and would love to keep the volume of changes here as low as possible.
We'll update RuboCop, and adjust/enforce the rules later on. Not to say that we need to fix that `rubocop` works differently locally and on CI.

pirj reviewed (commented):
Looks great!
A couple of cosmetic notes.

pirj commented:
Most build failures seem to be due to `"".dup` => `+""` change.

```
lib/rspec/core/example.rb:246:7: C: Metrics/MethodLength: Method has too many lines. [39/37]
      def run(example_group_instance, reporter) ...
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Please don't hesitate to relax `Metrics/MethodLength` to tolerate 39 lines.

odinhb commented on features/metadata/current_scope.feature:
```diff
@@ -0,0 +1,84 @@
+        end
+      end
+
+      RSpec.describe "RSpec::current_scope" do
```
Ah it's hard to kill habits. I like the `::`/`#` notation so much better for some reason.

pirj commented:
Ruby 2.7 failure seems unrelated:
```
An error occurred while loading spec_helper.
Failure/Error: require 'aruba/api'

TypeError:
  unable to resolve type 'intptr_t'
# ./vendor/bundle/ruby/2.7.0/gems/ffi-1.12.2-x64-mingw32/lib/ffi/types.rb:69:in `find_type'
```

So does a `ruby-head` one with RuboCop/psych arity mismatch:
```
bin/rubocop lib
wrong number of arguments (given 5, expected 1)
/home/runner/.rubies/ruby-head/lib/ruby/3.1.0/psych.rb:323:in `safe_load'
```

Otherwise looks green. @JonRowe, what do you think?

JonRowe commented on Changelog.md:
```diff
@@ -6,6 +6,7 @@ Enhancements:
 * Improve pluralisation of words ending with `s` (like process). (Joshua Pinter, #2779)
 * Add ordering by file modification time (most recent first). (Matheus Richard, #2778)
 * Add `to_s` to reserved names for #let and #subject. (Nick Flückiger, #2886)
+* Introduce `RSpec.current_scope`. (@odinhb, #2895)
```
```suggestion
* Introduce `RSpec.current_scope` to expose the current scope in which
  RSpec is executing. e.g. `:before_example_hook`, `:example` etc. (@odinhb, #2895)
```

JonRowe commented on features/metadata/current_scope.feature:
```diff
@@ -0,0 +1,84 @@
+Feature: RSpec provides the current_scope as RSpec::current_scope
```
```suggestion
Feature: RSpec provides the current_scope as RSpec.current_scope
```

JonRowe commented on lib/rspec/core.rb:
```diff
@@ -129,6 +129,31 @@ def self.current_example=(example)
+  #
+  # Returns `:example` inside it/example blocks
+  #
+  # Returns `nil` after your suite and all hooks are done
```
It should return `:suite` afterwards as well

odinhb replied:
Thought a bit about this. Returning `nil` after the suite is finished is an arbitrary quirk that I can't really justify for any reason.

Also If somehow you ran two suites back to back, it would return `:suite` before running the first one, `nil` in-between them (which is after the first and before the second) and `nil` again after the second.

Apologies for the arbitrariness of implementation.

pirj replied:
What comes to mind is `bisect` and a custom extension that retries failed examples.

Wondering **if** those extensions would need to distinguish pre-suite from post-suite.
Or calling it from a custom formatter. But the formatter would already know if they implement a `stop`/`close` event handler.

I suggest keeping it consistent for now. If it turns out some extension needs this information, we'll figure something out. Returning `:suite` is more elegant, and consistent, and the whole `current_scope` pretty much covers what we needed for `test-prof` and `rspec-rails`.

JonRowe commented on features/metadata/current_scope.feature:
```diff
@@ -0,0 +1,84 @@
+  Scenario: Detecting the current scope
+    Given a file named "current_scope_spec.rb" with:
+      """ruby
+      unless RSpec.current_scope == :suite
+        raise "bad current scope: #{RSpec.current_scope.inspect}"
+      end
```
I don't think this is necessary

```suggestion
```

pirj replied:
That seems to be the only way to test that it's `:suite` at this stage. It's kind of a weird "test" though.

odinhb replied:
It gave me reasonable feedback that things were not working correctly while working on it.

I'm not sure how else to test it.

Similarly I don't know how you would test what value it has after the suite is finished.

pirj replied:
> what value it has after the suite is finished

Will `at_exit { exit(1) unless RSpec.current_scope.nil? }` work?

odinhb replied:
> Will at_exit { exit(1) unless RSpec.current_scope.nil? } work?

Nice. Didn't know I could do that.

odinhb commented:
Is there anything else you want me to do?

Should I fixup the `fixup! Adjust feature title` commit into `Update features/metadata/current_scope.feature`?

pirj commented:
@odinhb No, it's all good. Thanks a lot for your contribution. 🙌 

@JonRowe Do you want to take another look?

pirj commented:
@odinhb CI is green. Would you like to send a PR against `rspec-rails`'s `main` that would make use of this new method?
`main` tracks the ongoing development of 5.x of `rspec-rails`, and it's set to depend on `4.x` `rspec-core` that would include this new method.
I suggest you to tweak [those dependency declarations](https://github.com/rspec/rspec-rails/blob/main/Gemfile-rspec-dependencies) to point to this PR's branch.

That might be the last drop needed to convince @JonRowe 

Good job!

odinhb commented:
> @odinhb CI is green. Would you like to send a PR against `rspec-rails`'s `main` that would make use of this new method?

I can do that. I can't start until after the weekend however.

pirj commented:
> until after the weekend however

This is fine.
I was going to wrap up a few remaining things for 3.99 this weekend, and it may take a while to code review and bring them to a desired shape. There's still time before 4.0 is cut.

pirj reviewed (approved):
Thanks!

@JonRowe This is the only open PR I can think of that it makes sense to merge before we proceed with 4.0 release plan.

JonRowe commented on lib/rspec/core/example_group.rb:
```diff
@@ -537,9 +537,15 @@ def self.store_before_context_ivars(example_group_instance)
         end
       end
 
+      # @deprecated use `RSpec.current_scope` instead
```
```suggestion
```

JonRowe replied:
Its not necessary to deprecate this here

pirj replied:
Cherry-picked deprecation to https://github.com/rspec/rspec-core/pull/2880

JonRowe commented on lib/rspec/core/example_group.rb:
```diff
@@ -537,9 +537,15 @@ def self.store_before_context_ivars(example_group_instance)
       # Returns true if a `before(:context)` or `after(:context)`
       # hook is currently executing.
       def self.currently_executing_a_context_hook?
+        RSpec.deprecate(
+          "currently_executing_a_context_hook",
+          :replacement => "RSpec.current_scope"
+        )
+
```
```suggestion
```

JonRowe commented on features/metadata/current_scope.feature:
```diff
@@ -0,0 +1,88 @@
+  Scenario: Detecting the current scope
+    Given a file named "current_scope_spec.rb" with:
+      """ruby
+      unless RSpec.current_scope == :suite
+        raise "bad current scope: #{RSpec.current_scope.inspect}"
+      end
+
+      at_exit do
+        exit(1) unless RSpec.current_scope == :suite
+      end
```
```suggestion
      # Outside of the test lifecycle, the current scope is `:suite`
      exit(1)  unless RSpec.current_scope == :suite
      
      at_exit do
        exit(1) unless RSpec.current_scope == :suite
      end
```

JonRowe reviewed (changes requested):
Some documentation changes mainly, and a request to remove a deprecation

pirj commented:
All green. Since there were no objections to the approach, and all amendments were considered, I'll go ahead and merge.
I intend to cherry-pick this to `4-0-dev` and also remove `currently_executing_a_context_hook?`. For 3.99 a deprecation warning [has been added](https://github.com/rspec/rspec-core/pull/2880/commits).

@odinhb Thanks for the contribution!
